docker/prepare_proxy.sh

```diff
@@ -34,26 +40,53 @@ while getopts ":p:u:h" opt; do
   esac
 done
 
-if [[ -z "${ISTIO_PROXY_PORT-}" ]] || [[ -z "${ISTIO_PROXY_UID-}" ]]; then
+if [[ -z "${ENVOY_PORT-}" ]] || [[ -z "${ENVOY_UID-}" ]]; then
   echo "Please set both -p and -u parameters"
   usage
   exit 1
 fi
 
-# 1. External traffic is redirected to proxy port
-iptables -t nat -A PREROUTING -p tcp -j REDIRECT --to-port ${ISTIO_PROXY_PORT}
-
-# 2. Locally generated traffic sent on loopback interface (because traffic was
-# routed locally) and destination IP is not explicitly the loopback IP
-# (e.g. 172.17.0.2 and not 127.0.0.1) is redirected back to proxy port.
-# This include any proxy generated traffic which is necessary to make
-# (app => proxy => proxy => app) work when two apps are same pod.
-iptables -t nat -A OUTPUT -p tcp -o lo -j REDIRECT \
-  ! -d 127.0.0.1/32 --to-port ${ISTIO_PROXY_PORT}
-
-# 3. Locally generated traffic not sent explicitly from loopback IP
-# (i.e. proxy aware) or not from the proxy itself is redirected proxy port.
-iptables -t nat -A OUTPUT -p tcp -j REDIRECT ! -s 127.0.0.1/32 \
-  --to-port ${ISTIO_PROXY_PORT} -m owner '!' --uid-owner ${ISTIO_PROXY_UID}
+# Create a new chain for redirecting inbound and outbound traffic to
+# the common Envoy port.
+iptables -t nat -N ISTIO_REDIRECT                                             -m comment --comment "istio/redirect-common-chain"
+iptables -t nat -A ISTIO_REDIRECT -p tcp -j REDIRECT --to-ports ${ENVOY_PORT} -m comment --comment "istio/redirect-to-envoy-port"
+
+# Redirect all inbound traffic to Envoy.
+iptables -t nat -A PREROUTING -j ISTIO_REDIRECT                               -m comment --comment "istio/install-istio-prerouting"
+
+# Create a new chain for selectively redirecting outbound packets to
+# Envoy.
+iptables -t nat -N ISTIO_OUTPUT                                               -m comment --comment "istio/common-output-chain"
+
+# Jump to the ISTIO_OUTPUT chain from OUTPUT chain for all tcp
+# traffic. '-j RETURN' bypasses Envoy and '-j ISTIO_REDIRECT'
+# redirects to Envoy.
+iptables -t nat -A OUTPUT -p tcp -j ISTIO_OUTPUT                              -m comment --comment "istio/install-istio-output"
+
+# Redirect app calls to back itself via Envoy when using the service VIP or endpoint
+# address, e.g. appN => Envoy (client) => Envoy (server) => appN.
+iptables -t nat -A ISTIO_OUTPUT -o lo ! -d 127.0.0.1/32 -j ISTIO_REDIRECT     -m comment --comment "istio/redirect-implicit-loopback"
+
+# Avoid infinite loops. Don't redirect Envoy traffic directly back to
+# Envoy for non-loopback traffic.
+iptables -t nat -A ISTIO_OUTPUT -m owner --uid-owner ${ENVOY_UID} -j RETURN   -m comment --comment "istio/bypass-envoy"
+
+# Skip redirection for Envoy-aware applications and
+# container-to-container traffic both of which explicitly use
+# localhost.
+iptables -t nat -A ISTIO_OUTPUT -d 127.0.0.1/32 -j RETURN                     -m comment --comment "istio/bypass-explicit-loopback"
```

Comment:
Agreed. Most of the setup/teardown code can be shared with the other integration tests. Maybe after release we can split that out into its own test package and have individual tests consume that that?